qubes/ext/core_features.py

```diff
@@ -34,6 +36,32 @@ async def qubes_features_request(self, vm, event, untrusted_features):
                 'Ignoring qubes.NotifyTools for template-based VM')
             return
 
+        if "os-distribution" in untrusted_features:
+            # entry point already validats values for safe characters
+            vm.features["os-distribution"] = untrusted_features["os-distribution"]
+        if "os-version" in untrusted_features:
+            # no letters in versions please
+            safe_set = string.digits + ".-"
+            untrusted_version = untrusted_features["os-version"]
+            if all(c in safe_set for c in untrusted_version):
+                vm.features["os-version"] = untrusted_version
+            else:
+                # safe to log the value as passed preliminary filtering already
+                vm.log.warning("Invalid 'os-version' value '%s', only digits and _ or . are allowed", untrusted_version)
+        if "os-eol" in untrusted_features:
+            untrusted_eol = untrusted_features["os-eol"]
+            valid = False
+            if re.match(r"\d{4}-\d{2}-\d{2}", untrusted_eol):
```

Comment:
```suggestion
            if re.match(r"\A\d{4}-\d{2}-\d{2}\Z", untrusted_eol):
```

don’t allow trailing junk